library/alloc/src/sync.rs

```diff
@@ -328,6 +328,79 @@ impl<T> Arc<T> {
         Self::from_inner(Box::leak(x).into())
     }
 
+    /// Constructs a new `Arc<T>` using a weak reference to itself. Attempting
+  /// to upgrade the weak reference before this function returns will result
+  /// in a `None` value. However, the weak reference may be cloned freely and
+  /// stored for use at a later time.
+  ///
+  /// # Examples
+  /// ```
+  /// #![feature(arc_new_cyclic)]
+  /// #![allow(dead_code)]
+  ///
+  /// use std::sync::{Arc, Weak};
+  ///
+  /// struct Foo {
+  ///     me: Weak<Foo>,
+  /// }
+  ///
+  /// let foo = Arc::new_cyclic(|me| Foo {
+  ///     me: me.clone(),
+  /// });
+  /// ```
```

Comment:
What is the use case for this example? I don't understand by just looking at the example, maybe we should provide a more realistic use case?